### ISSUE TITLE
Error with dynamic fields when validating form

### ISSUE THREAD

**Issue by arielger** [jaredpalmer/formik#1268]
## 🐛 Bug report

First of all, thanks for your hard work in this library! 🙌🕺

### Current Behavior

When creating a form with Fields that are rendered depending on the value of other field the values of the form won't update according to those dynamic fields.

If I remove a dynamic field from the form, the value will still be in the values object.

If I render a dynamic field with no associated initial value and then submit the form without changing the value of that input, it won't show any related error. This is because the new Field value is not in `values` until you change the value for the first time and when submitting Formik is only setting the already present values to `touched: true` (https://github.com/jaredpalmer/formik/blob/master/src/Formik.tsx#L435).

### Expected behavior

I would expect that:
- When you remove a Field from the form, the related value is removed from `values`.
- When rendering a new Field it should add its initial value to `values` so when submitting you can see the related errors for that Field (maybe adding an `initialValue` prop for the Field component and setting that in the values object in `componentDidMount`). 

What I'm doing right now to fix this is running a custom `onChange` function in the first input and according to the value, using `setValues` to remove the old key of the dependent input and adding a default for the new key.

### Reproducible example

https://codesandbox.io/s/jn5z79w7ww

### Suggested solution(s)

- In [`unregisterField`](https://github.com/jaredpalmer/formik/blob/0e459bce5f88bc60681f81ffe13df81982a723ab/src/Formik.tsx#L86) we could remove the related value from `values`.
- Add `initialValue` prop for Field component and default to that value if there is no value in `initialValues`.

If you think this makes sense let me know and I can work on a PR.

### Your environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Formik           | 1.4.2
| React            | 16.5.0
| TypeScript       | - 
| Browser          | Google Chrome
| npm/Yarn         | Reproduced in CodeSandbox
| Operating System | MacOS

**Comment by jaredpalmer** [comment]
I agree that this is wonky right now, but relying on on render/mount is a dangerous path to go down. Just yesterday, we were working internally on v2 and came up some different registration techniques with `useRef` in addition to `registerField`. I think the other thing to remember is that not everyone uses `<Field>` to begin with, which complicates things.

I guess my point is that it's unclear that removing the value is indeed desired on unmount. What if you're building a form like https://www.typeform.com/ where there is like one input per page? What if you are animating inputs and gently moving your user down the form? What if you are using flip animations to transition from input to input? I feel you that is is a bit strange, and maybe we want a `removeOnUnMount` prop for `Field`, but I don't think deletion should be the default.  Food for thought. Maybe I have it wrong.

**Comment by jaredpalmer** [comment]
Oh I forgot to mention. There are multiple "issues" you brought up in your initial message. 

As it relates to default value for `<Field>`, the reason it is like this is because that's how React works with normal inputs...and `<Field>` is a convenience component that tries to cut down on boilerplate without adding magic.

**Comment by jaredpalmer** [comment]
`<Field initialValue>` is a decent idea that I think solves for this. However, I'm unclear as to whether it should modify the whole form's `initialValues` and what should happen with that `initialValue` when a user calls `resetForm`.

**Comment by kimochg** [comment]
@jaredpalmer I run into a case in which the `initialValues` is a little hard to decide when rendering top-level  `Formik`, it would be nice to have sth like `<Field initialValue>` or expose sth like `setFieldInitialValue` to make field-level initial value setting possible.

**Comment by drock07** [comment]
Are there any recommendations on how to handle the first issue:

> If I remove a dynamic field from the form, the value will still be in the values object.

Currently, in my validation function, I'm creating errors for fields that are no longer in the form because their values are still in Formik's state. This is preventing the form from submitting. 

Is there anything on the roadmap about adding an option to remove a value when it is unregistered?

**Comment by arielger** [comment]
Not sure if it's an ideal solution but you can use [`setValues`](https://jaredpalmer.com/formik/docs/api/formik#setvalues-fields-field-string-any-void) to remove the value associated with the unmounted component.

**Comment by Danziger** [comment]
> What if you're building a form like https://www.typeform.com/ where there is like one input per page? 

I'm not sure that's representative of the forms most people would build, so maybe the default should be removing the value if the field goes away and there should be an option to change this behaviour to keep the values.

**Comment by vaneavasco** [comment]
Any updates?

**Comment by wemakeweb** [comment]
is there any solution?

**Comment by jrea** [comment]
```
  const revalidate = useCallback(async () => {
    await validateForm();
  }, [validateForm]);

  useEffect(() => {
    setHideField(hideField);
    if (hideField === false) {
      revalidate();
      setValidatedOnce(true);
    }
  }, [hideField, revalidate]);

  useEffect(() => {
    if (validatedOnce === true) {
      revalidate();
      setValidatedOnce(false);
    }
  }, [revalidate, validatedOnce]);
```
In my specific case, there's a whole section of the form that gets removed, so doing it via `setField` made me dead inside.
